clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -219,31 +410,58 @@ class device_vector {
   T *real_begin() { return _storage; }
   const T *real_begin() const { return _storage; }
   void swap(device_vector &v) {
-    ::std::swap(_size, v._size);
-    ::std::swap(_capacity, v._capacity);
-    ::std::swap(_storage, v._storage);
-    ::std::swap(_alloc, v._alloc);
+    if constexpr (::std::allocator_traits<
+                      Allocator>::propagate_on_container_swap::value) {
+      ::std::swap(_alloc, v._alloc);
+      ::std::swap(_size, v._size);
+      ::std::swap(_capacity, v._capacity);
+      ::std::swap(_storage, v._storage);
+    } else {
+      // swap all elements up to the minimum size between the two vectors
+      size_type min_size = ::std::min(size(), v.size());
+      for (size_type i = 0; i < min_size; i++) {
+        ::std::swap(_storage[i], v._storage[i]);
+      }
+      // then copy the elements beyond the end of the smaller list, and resize
+      if (size() > v.size()) {
+        size_type tmp_v_size = v.size();
+        v.resize(size());
```

Comment:
Same as above.

Reply:
Thanks for catching this, it needs a bit of work.  We probably need special casing for different capacity vs size situations.  We also may want a special swapping parallel for in order to avoid bringing the data to the host to swap it, and to extra temporary allocation requirements.
Uninitialized fill makes sense for the excess.

Reply:
fixed, thanks.